### PR TITLE
#24: improve platform compatibility by changing default random seed

### PR DESCRIPTION
**Pull request by jeking3**
from uuid internal code which was not friendly to all platforms, to boost::random::random_device which is
I'm submitting this PR for comments.

@eldiener this requires a new dependency for some use cases; folks may require linking against boost::random if they are using boost::uuid and relying on the default random generator constructor whereas before the code was platform specific and not always compatible.

Summary:

1. Moved seed_rng into detail/
2. Removed all platform-specific logic in seed_rng and changed it to leverage boost::random::random_device.
3. Updated documentation and comments appropriate to change.

Should consider moving seed_rng into boost::random at some point to make it easier to use seed_seq, or fastly simplify boost::random seeding of PRNGs by allowing each PRNG to take a boost::random::random_device as an argument to seed().

### CODE REVIEW

**Comment by eldiener** [comment]
Why should just changing the order of the header files in random_generator.hpp ever matter ?

**Comment by jeking3** [comment]
While header order should not (and in this case does not) matter, I prefer
to have headers ordered alphabetically with the exception of the header for
the module it is implementing which belongs first.

File: include/boost/uuid/detail/seed_rng.hpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+namespace boost {
+namespace uuids {
+namespace detail {
+
+// seed() seeds a random number generator with good seed values
+template <typename UniformRandomNumberGenerator>
+inline void seed(UniformRandomNumberGenerator& rng)
+{
+    BOOST_CONSTEXPR_OR_CONST size_t entropy_len = 16; // 512 bytes of entropy
```
Comment (jeking3) [inline include/boost/uuid/detail/seed_rng.hpp]: Should boost/config.hpp be added as an include to satisfy BOOST_CONSTEXPR_OR_CONST usage here?

**Comment by strentler** [comment]
@jeking3 : i have applied your patch. It does the job for UWP now.  Have tested this with SDK version 10.0.15063.0. I have also checked if it goes through the "Windows App Cert Kit".  It does.
Thank you!

**Comment by jeking3** [comment]
@eldiener or anyone else on the CTP team able to do a code review?

File: include/boost/uuid/detail/seed_rng.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+namespace uuids {
+namespace detail {
+
+// seed() seeds a random number generator with good seed values
+template <typename UniformRandomNumberGenerator>
+inline void seed(UniformRandomNumberGenerator& rng)
+{
+    BOOST_STATIC_CONSTEXPR size_t entropy_len = 32; // 256 bits of entropy
+    boost::random::random_device rd;
+    std::vector<unsigned int> entropy(entropy_len);
```
Comment (Lastique) [inline include/boost/uuid/detail/seed_rng.hpp]: This should be an array if the size is known in compile time (it is in your case).

File: include/boost/uuid/detail/seed_rng.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+namespace boost {
+namespace uuids {
+namespace detail {
+
+// seed() seeds a random number generator with good seed values
+template <typename UniformRandomNumberGenerator>
+inline void seed(UniformRandomNumberGenerator& rng)
+{
+    BOOST_STATIC_CONSTEXPR size_t entropy_len = 32; // 256 bits of entropy
```
Comment (Lastique) [inline include/boost/uuid/detail/seed_rng.hpp]: `std::size_t`. Also, `#include <cstddef>` is missing.

File: include/boost/uuid/detail/seed_rng.hpp [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    boost::random::random_device rd;
+    std::vector<unsigned int> entropy(entropy_len);
+    std::generate(entropy.begin(), entropy.end(), std::ref(rd));
+    boost::random::seed_seq seq(entropy.begin(), entropy.end());
+    rng.seed(seq);
+}
+
+// random_device itself does not / can not be seeded
+template <>
+inline void seed<boost::random::random_device>(boost::random::random_device&) {}
```
Comment (Lastique) [inline include/boost/uuid/detail/seed_rng.hpp]: This should probably be an overload instead of specialization.

**Comment by Lastique** [comment]
I would rather keep `seed_rng` so that it is possible to generate UUIDs without linking with Boost.Random. `seed_rng` uses system sources for getting random data that are ultimately used by Boost.Random, so there's little reason to bring in that dependency. I would rather have Boost.Random-based version separate, if possible, and without breaking backward compatibility.

**Comment by jeking3** [comment]
@Lastique my rationale for eliminating the platform-specific code in seed_rng was to avoid situations such as that described in #24.  Given there is already a documented, supported, and highly portable seed generator in boost::random::random_device, it makes less sense to me to have a similar method in uuid with platform specific code that needs to be maintained.  Could you explain further how you believe the change I made is not backwards compatible?  seed_rng is in the detail namespace and it is my understanding that these do not need to maintain backwards compatibility.  

As with all projects, I am looking to minimize the maintenance cost of the library by leveraging that which already exists and likely does a far better job than the code that I removed.  It should be intuitive from a user perspective that if they are using random_generator to make randomized UUIDs, that they would need to link to boost::random.

Thank you for your other code review comments, I will make changes to address them.

**Comment by Lastique** [comment]
On 09/07/17 20:00, James E. King, III wrote:
> my rationale for eliminating the 
> platform-specific code in seed_rng was to avoid situations such as that 
> described in #24 <https://github.com/boostorg/uuid/issues/24>. Given 
> there is already a documented, supported, and highly portable seed 
> generator in boost::random::random_device, it makes less sense to me to 
> have a similar method in uuid with platform specific code that needs to 
> be maintained.

I don't think the cost of maintenance here is significant enough when 
compared to having Boost.UUID a header-only library. In some projects, 
header-only property is a deal breaker.

The #24 needs to be investigated. I'm not familiar with the UWP 
specifics, but I'm sure there must be an API for accessing system RNG.

> Could you explain further how you believe the change I 
> made is not backwards compatible? seed_rng is in the detail namespace 
> and it is my understanding that these do not need to maintain backwards 
> compatibility.

Having to link with Boost.Random is a breaking change.

> It should be intuitive from 
> a user perspective that if they are using random_generator to make 
> randomized UUIDs, that they would need to link to boost::random.

No, I'm sorry, that doesn't follow. Boost.Random is a separate library 
and Boost.UUID is not obliged to use it if there is enough reason not 
to. IMHO, header-only property is important enough to avoid the linking 
dependency on Boost.Random, especially since the seeding code is already 
written and works on most platforms (except UWP, that is). Adding 
support for another platform to the existing implementation seems more 
preferable to me.

**Comment by jeking3** [comment]
@Lastique your argument for header-only is convincing enough for me to reconsider.  I do wonder though if seed_rng belongs in boost::random more than in uuid, since there will be subject-matter experts there who will be able to deal with the issue of platform compatibility should it pop up again in the future.  For the time being I will close out this change and focus on fixing the specific bug reported.

**Comment by Lastique** [comment]
> I do wonder though if seed_rng belongs in boost::random more than in uuid

This is probably true. `seed_rng` wouldn't be needed if `boost::random_device` was header-only.

**Pull request closed without merging**